### PR TITLE
imeplement plc4x

### PR DESCRIPTION
**Pull request by MrLeea-13155bc**
**What this PR does / why we need it**:
imeplement deviceshifu-plc4x
**Will this PR make the community happier**? 
yes
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**How is this PR tested**
- [ ] unit test
- [ ] e2e test
- [x] other (please specify)
now with alittle test
**Special notes for your reviewer**:

Design for this PR: #270 

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by saiyan86** [comment]
#270

**Review by BtXin — APPROVED** [review]
LGTM

File: pkg/k8s/api/v1alpha1/edgedevice_types.go [hunk 1]
```diff
@@ -45,6 +45,24 @@ type OPCUASetting struct {
...
+
+const (
+	Plc4xProtocolS7          Plc4xProtocol = "s7"
+	Plc4xProtocolADS         Plc4xProtocol = "ads"
+	Plc4xProtocolBACnet      Plc4xProtocol = "bacnet"
+	Plc4xProtocolCBus        Plc4xProtocol = "cbus"
+	Plc4xProtocolEip         Plc4xProtocol = "eip"
+	Plc4xProtocolKnx         Plc4xProtocol = "knx"
+	Plc4xProtocolModbusAscii Plc4xProtocol = "modbus-ascii"
+	Plc4xProtocolModbusRTU   Plc4xProtocol = "modbus-rut"
```
Comment (tomqin93) [inline pkg/k8s/api/v1alpha1/edgedevice_types.go:62]: `modbus-rut` => `modbus-rtu`

**Review by tomqin93 — COMMENTED** [review]
@MrLeea-13155bc  added some comment, please take a look

File: pkg/deviceshifu/utils/utils.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		paramStr = url[0]
+	} else {
+		paramStr = url[1]
+	}
+
+	params := strings.Split(paramStr, "&")
+
+	result := make(map[string]string, len(params))
+
+	for _, item := range params {
```
[the remark this answers is not in the document]
Comment (MrLeea-13155bc) [reply pkg/deviceshifu/utils/utils.go:26]: `req.URL.Query().Get()` can only get a value for which we know the key.

File: pkg/deviceshifu/deviceshifuplc4x/deviceshifuplc4x.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		if err != nil {
+			klog.Errorf("Error when build request by params, error: %v", err)
+			http.Error(w, "Error when build request with params, error: "+err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		var resultData model.PlcWriteRequestResult
+		select {
+		case resultData = <-plc4xRequest.Execute():
+		case <-time.After(time.Second):
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifuplc4x/deviceshifuplc4x.go:123]: should we make this timeout configurable? same for the read timeout

**Review by tomqin93 — COMMENTED** [review]
@MrLeea-13155bc added 2 comments, please take a look

**Review by tomqin93 — APPROVED** [review]
LGTM

**Pull request merged**